### PR TITLE
Added wiki_zh dataset preprocessing script

### PR DESCRIPTION
**Pull request by qijiaxing**
Added a script preprocessing the wiki_zh dataset as an example to show BERT pretraining on a Chinese dataset. Updated the `bert_pretraining.py` example a bit as well, allowing users to choose `NemoBertTokenizer` for training on Chinese.

### CODE REVIEW

**Comment by okuchaiev** [comment]
PEP8 issue:

+ pycodestyle . --exclude=./scripts/get_librispeech_data.py,./scripts/process_beam_dump.py,./tests/other/jasper.py,./tests/other/jasper_zero_dl.py,./collections/nemo_nlp/nemo_nlp/utils/metrics/sacrebleu.py,./collections/nemo_nlp/nemo_nlp/utils/metrics/fairseq_tokenizer.py,./nemo/setup.py,./docs/sources/source/conf.py,./docs/sources/source/tutorials/infer.py,./docs/sources/source/tutorials/test.py,./collections/nemo_nlp/build

./examples/nlp/scripts/process_wiki_zh.py:139:1: E305 expected 2 blank lines after class or function definition, found 1

script returned exit code 1

File: examples/nlp/bert_pretraining.py [hunk 1]
```diff
@@ -54,17 +57,22 @@
...
                                     'train.txt')
 
-tokenizer = nemo_nlp.SentencePieceTokenizer(
-    model_path=data_desc.tokenizer_model)
-tokenizer.add_special_tokens(special_tokens)
+if args.tokenizer == "sentence-piece":
+    tokenizer = nemo_nlp.SentencePieceTokenizer(
+        model_path=data_desc.tokenizer_model)
+    tokenizer.add_special_tokens(special_tokens)
+else:
```
Comment (redoctopus) [inline examples/nlp/bert_pretraining.py]: Please make this explicit (case specifically for "nemo-bert") and have the else case raise an error.
We want to make sure that if the user mistypes `sentence-piec` or something that we won't silently fall through to a different default than they expect.

Comment (qijiaxing) [reply examples/nlp/bert_pretraining.py]: Thanks! Revised the code accordingly.

File: examples/nlp/scripts/process_wiki_zh.py [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+                if w in vocab:
+                    vocab[w] += 1
+                else:
+                    vocab[w] = 1
+
+    # Remove rare words
+    for w in vocab.keys():
+        if vocab[w] < min_frequency:
+            del vocab[w]
+            print("Remove rare word {} from vocab".format(w))
```
Comment (chiphuyen) [inline examples/nlp/scripts/process_wiki_zh.py]: alternative if you don't have to care about extra memory footprint:
new_vocab = {w: vocab[w] for w in vocab if vocab[3] >= min_frequency}

File: examples/nlp/scripts/process_wiki_zh.py [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+
+    # Include special symbols and write to file
+    vocab = list(special_symbols) + sorted(vocab.keys())
```
Comment (chiphuyen) [inline examples/nlp/scripts/process_wiki_zh.py]: special_symbols can be passed in as a list so you don't have to convert it later

File: examples/nlp/scripts/process_wiki_zh.py [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+    with open(vocab_file, 'w') as f:
+        for w in vocab:
+            f.write(w + '\n')
+    print(f"Created vocabulary of size {len(vocab)}")
```
Comment (chiphuyen) [inline examples/nlp/scripts/process_wiki_zh.py]: use logger

File: examples/nlp/scripts/process_wiki_zh.py [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+    with open(filename, encoding='utf-8') as f:
+        documents = f.readlines()
+
+    r = re.compile(regex)
+    # Each line is a full article
+    for i, line in enumerate(documents):
+        text = json.loads(line)['text']
+        text = text.replace('\\n', '\n').strip()
+        text = r.findall(text)
+        text = ' '.join([c for c in text])  # separate char by space
```
Comment (chiphuyen) [inline examples/nlp/scripts/process_wiki_zh.py]: this means you're iterating over the text twice (one to split and one to add space). depends on how long your text is, you might want to iterate only once.

File: examples/nlp/bert_pretraining.py [hunk 6]
```diff
@@ -54,17 +57,27 @@
...
+        model_path=data_desc.tokenizer_model)
+    tokenizer.add_special_tokens(special_tokens)
+elif args.tokenizer == "nemo-bert":
+    nf.logger.info("To use NemoBertTokenizer.")
+    vocab_file = os.path.join(args.data_dir, 'vocab.txt')
+    # To train on a Chinese dataset, use NemoBertTokenizer
+    tokenizer = nemo_nlp.NemoBertTokenizer(vocab_file=vocab_file)
+else:
+    raise ValueError(f"{args.tokenizer} Tokenizer not supported! "
+                     "Please use sentence-piece or nemo-bert")
```
Comment (chiphuyen) [inline examples/nlp/bert_pretraining.py]: Please add your tokenizer or use sentence-piece or nemo-bert.

File: examples/nlp/scripts/process_wiki_zh.py [hunk 7]
```diff
@@ -0,0 +1,155 @@
...
+def read_files(data_dir, regex, max_dirs=-1):
+    """Read text in all sub-directories, filter by regex and return lines"""
+    executor = ProcessPoolExecutor(max_workers=4)
+
+    tasks = []
+    sub_dirs = [name for name in os.listdir(data_dir)
+                if os.path.isdir(os.path.join(data_dir, name))]
+    for sub_dir in sub_dirs[:max_dirs]:
+        p_dir = os.path.join(data_dir, sub_dir)
+        print("To process files in dir: "+p_dir)
```
Comment (chiphuyen) [inline examples/nlp/scripts/process_wiki_zh.py]: use f-string & logger

File: examples/nlp/scripts/process_wiki_zh.py [hunk 8]
```diff
@@ -0,0 +1,155 @@
...
+        files = os.listdir(p_dir)
```
Comment (chiphuyen) [inline examples/nlp/scripts/process_wiki_zh.py]: if you use glob.glob then you don't have to convert the files to fullpaths (glob returns fullpath)

File: examples/nlp/scripts/process_wiki_zh.py [hunk 9]
```diff
@@ -0,0 +1,155 @@
...
+        for f in files:
+            fullpath = os.path.join(data_dir, sub_dir, f)
+            tasks.append(executor.submit(partial(read, fullpath, regex)))
+    print('Preprocessing texts, please wait...', flush=True)
+
+    # Collect text
+    lines = []
+    for t in tasks:
+        for doc in t.result():
+            splitted = re.split(r'[\n]', doc)
```
Comment (chiphuyen) [inline examples/nlp/scripts/process_wiki_zh.py]: doc.split('\n') does the same thing

File: examples/nlp/scripts/process_wiki_zh.py [hunk 10]
```diff
@@ -0,0 +1,155 @@
...
+            lines += splitted
+
+    return lines
+
+
+def save(output_dir, lines, train_p=0.95):
+    """Save lines into two files within output dir: train.txt and val.txt"""
```
Comment (chiphuyen) [inline examples/nlp/scripts/process_wiki_zh.py]: add doc about train_p. you might want to name it train_ratio for consistency with other processors

File: examples/nlp/scripts/process_wiki_zh.py [hunk 11]
```diff
@@ -0,0 +1,155 @@
...
+    # Take large portino as training set
+    train_file = os.path.join(output_dir, 'train.txt')
+    num_train_doc = int(train_p * len(lines))
+    with open(train_file, 'w', encoding='utf-8') as f:
+        for l in lines[:num_train_doc]:
+            f.write(l+'\n')
```
Comment (chiphuyen) [inline examples/nlp/scripts/process_wiki_zh.py]: spaces around +

File: examples/nlp/scripts/process_wiki_zh.py [hunk 12]
```diff
@@ -0,0 +1,155 @@
...
+
+    # Take remainings as validation set
+    val_file = os.path.join(output_dir, 'valid.txt')
+    with open(val_file, 'w', encoding='utf-8') as f:
+        for l in lines[num_train_doc:]:
+            f.write(l+'\n')
+
+    print("Finished processing.")
+    print("Processed text saved to {} and {}".format(train_file, val_file))
```
Comment (chiphuyen) [inline examples/nlp/scripts/process_wiki_zh.py]: f-string + logger

File: examples/nlp/scripts/process_wiki_zh.py (hunk 2 above)
Comment (qijiaxing) [reply examples/nlp/scripts/process_wiki_zh.py]: Thanks! Revised this.

File: examples/nlp/scripts/process_wiki_zh.py (hunk 3 above)
Comment (qijiaxing) [reply examples/nlp/scripts/process_wiki_zh.py]: Thanks! Revised this.

File: examples/nlp/scripts/process_wiki_zh.py (hunk 4 above)
Comment (qijiaxing) [reply examples/nlp/scripts/process_wiki_zh.py]: We don't import nemo in this stand along pure python script, thus I didn't use logger.

File: examples/nlp/scripts/process_wiki_zh.py (hunk 8 above)
Comment (qijiaxing) [reply examples/nlp/scripts/process_wiki_zh.py]: Thanks! Improved this by using `glob`

File: examples/nlp/scripts/process_wiki_zh.py (hunk 9 above)
Comment (qijiaxing) [reply examples/nlp/scripts/process_wiki_zh.py]: Thanks! Revised accordingly.

File: examples/nlp/scripts/process_wiki_zh.py (hunk 7 above)
Comment (qijiaxing) [reply examples/nlp/scripts/process_wiki_zh.py]: Thanks! Used f-string, but not logger as we don't import nemo in this script.

File: examples/nlp/bert_pretraining.py (hunk 6 above)
Comment (qijiaxing) [reply examples/nlp/bert_pretraining.py]: Thanks! Revised accordingly.

File: examples/nlp/scripts/process_wiki_zh.py (hunk 10 above)
Comment (qijiaxing) [reply examples/nlp/scripts/process_wiki_zh.py]: Thanks! Renamed it `train_ratio` and revised doc.

File: examples/nlp/scripts/process_wiki_zh.py (hunk 11 above)
Comment (qijiaxing) [reply examples/nlp/scripts/process_wiki_zh.py]: Thanks! Revised this.

File: examples/nlp/scripts/process_wiki_zh.py (hunk 12 above)
Comment (qijiaxing) [reply examples/nlp/scripts/process_wiki_zh.py]: Thanks! Used f-string, but no logger as we don't import nemo in this script.

File: examples/nlp/scripts/process_wiki_zh.py (hunk 5 above)
Comment (qijiaxing) [reply examples/nlp/scripts/process_wiki_zh.py]: Thanks! Improved this by doing filtering and adding space in one iteration.

**Pull request merged**